src/numerics/petsc_matrix.C

```diff
@@ -1201,7 +1201,86 @@ void PetscMatrix<T>::add (const T a_in, const SparseMatrix<T> & X_in)
 }
 
 
+template <typename T>
+void PetscMatrix<T>::matrix_matrix_mult (SparseMatrix<T> & X_in, SparseMatrix<T> & Y_out, const MatReuse & Reuse)
+{
+  libmesh_assert (this->initialized());
+
+   // sanity check.
+   libmesh_assert_equal_to (this->n(), X_in.m());
+   libmesh_assert_equal_to (this->m(), Y_out.m());
+   libmesh_assert_equal_to (X_in.n(), Y_out.n());
+
+   const PetscMatrix<T> * X = cast_ptr<const PetscMatrix<T> *> (&X_in);
+   PetscMatrix<T> * Y = cast_ptr<PetscMatrix<T> *> (&Y_out);
+
+   libmesh_assert (X);
+   libmesh_assert (Y);
+
+   PetscErrorCode ierr=0;
+
+   // the matrix from which we copy the values has to be assembled/closed
+   libmesh_assert(X->closed());
+
+   semiparallel_only();
+
+   ierr = MatMatMult(_mat, X->_mat, Reuse, PETSC_DEFAULT, &Y->_mat);
+   LIBMESH_CHKERR(ierr);
+
+}
+
+template <typename T>
+void PetscMatrix<T>::add_sparse_matrix (const SparseMatrix<T> & spm,
+                        const std::map<numeric_index_type, numeric_index_type> & row_ltog,
+                        const std::map<numeric_index_type, numeric_index_type> & col_ltog,
+                        const T scalar)
+{
+  libmesh_assert_equal_to(spm.m(), row_ltog.size());
+  libmesh_assert_equal_to(spm.n(), col_ltog.size());
+  // make sure matrix has larger size than spm
+  libmesh_assert_greater_equal(this->m(), spm.m());
+  libmesh_assert_greater_equal(this->n(),spm.n());
 
+  if(!this->closed())
+    this->close();
+
+  PetscErrorCode ierr=0;
+
+  ierr=MatSetOption(_mat, MAT_NEW_NONZERO_ALLOCATION_ERR, PETSC_FALSE);
+  LIBMESH_CHKERR(ierr);
+
+  const PetscMatrix<T> * pscm=dynamic_cast<const PetscMatrix<T> *>(&spm);
+
+  PetscInt ncols=0;
+
+  const PetscInt *lcols;
+  const PetscScalar *vals;
+
+  for (auto ltog: row_ltog)
+  {
+    PetscInt grow[]={static_cast<PetscInt>(ltog.second)}; // global row index
+
+    ierr = MatGetRow(pscm->_mat, static_cast<PetscInt>(ltog.first), &ncols, &lcols, &vals);
+    LIBMESH_CHKERR(ierr);
+    // get global indices (gcols) from lcols, increment values = vals*scalar
+    PetscInt gcols[static_cast<std::size_t>(ncols)];
+    PetscScalar values[static_cast<std::size_t>(ncols)];
+    for (auto i=0; i<static_cast<std::size_t>(ncols); ++i)
```

Comment:
Since `ncols` is a `PetscInt`, just declare `i` to be a `PetscInt` as well. IIRC, the important thing is as far as efficiency is concerned is that the loop iteration counter type matches the loop end type, so no reason to involve the unsigned `std::size_t` type here.